ion/cmp_test.go

```diff
@@ -51,8 +51,28 @@ func (thisTimestamp ionTimestamp) eq(other ionEqual) bool {
 	return false
 }
 
-func cmpAnnotations(thisAnnotations, otherAnnotations []string) bool {
-	return reflect.DeepEqual(thisAnnotations, otherAnnotations)
+func cmpAnnotations(thisItem, otherItem ionItem) bool {
+	// Annotation sets are considered equal if the first annotation for each is $ion_symbol_table.
+	// eg. $ion_symbol_table::foo and $ion_symbol_table::bar::baz are considered equal.
```

Comment:
Why? Just for the purposes of this test suite, or is this from the Ion spec?

Reply:
The comments in `ion-tests/iontestdata/good/equivs/localSymbolTableWithAnnotations.ion` and `ion-tests/iontestdata/good/non-equivs/localSymbolTableWithAnnotations.ion` indicated this.

Reply:
These comments are [paraphrased from the spec](http://amzn.github.io/ion-docs/docs/symbols.html#annotating-local-symbol-tables); they clarify how the presence of extra annotations does or does not affect symbol table processing. I do not believe that they impact how annotation set equivalence should be evaluated.

-----

The [comment in `equivs/localSymbolTableWithAnnotations.ion`](https://github.com/amzn/ion-tests/blob/master/iontestdata/good/equivs/localSymbolTableWithAnnotations.ion#L1-L3) says: 

> // Local symbol table structs may have additional annotations as long as the $ion_symbol_table annotation comes first.
> // Additionally, values within the local symbols struct may be annotated.

The [comment in `non-equivs/localSymbolTableWithAnnotations.ion`](https://github.com/amzn/ion-tests/blob/034503caa56683f086edb1cc05a1b6207c1b18af/iontestdata/good/non-equivs/localSymbolTableWithAnnotations.ion#L1-L2) says:

> // If local symbol table structs have additional annotations, $ion_symbol_table must be the first. 
> // Otherwise, the struct is treated as a user value.

For an example, consider this struct. A reader would know to process it as a symbol table because it is annotated with `$ion_symbol_table`.
```js
$ion_symbol_table::{
    symbols: ["foo", "bar", "baz"]
}
```
You can add other annotations to that struct. If `$ion_symbol_table` is the first annotation, the struct will still be processed as a symbol table. Otherwise, `$ion_symbol_table` just becomes another annotation on a non-symbol-table user-level struct.
```js
// This is still a symbol table even though it has the 'quux' annotation
$ion_symbol_table::quux::{
    symbols: ["foo", "bar", "baz"]
}
// This is not a symbol table.
quux::$ion_symbol_table::{
    symbols: ["foo", "bar", "baz"]
}
```

-----

All of that said, I don't believe this statement from the PR is true:

> `$ion_symbol_table::foo` and `$ion_symbol_table::bar::baz` are considered equal

Reply:
In order to pass the equivalence tests, we need the following to be considered equal.

```js
$ion_symbol_table::{
    symbols:["foo", "bar"]
}
```
```js
$ion_symbol_table::annotated::{
    symbols:["foo", "bar"]
}
```
```js
$ion_symbol_table::annotated::{
    symbols:abc::["foo", def::"bar"]
}
```

So `$ion_symbol_table` and `$ion_symbol_table::annotated` need to pass the annotation comparison even though they are internally represented as `[]string{"$ion_symbol_table"}` and `[]string{"$ion_symbol_table", "annotated"}` respectively.
Likewise `symbols:["foo", "bar"]` and `symbols:abc::["foo", def::"bar"]` need to pass the comparison because they are inside a `$ion_symbol_table` annotated struct.

My solution is to bypass the full DeepEqual comparison between annotation sets whenever we're dealing with this `$ion_symbol_table` special case.

Do you have an issue with my solution or rather with my comments explaining the special case?

Reply:
> In order to pass the equivalence tests, we need the following to be considered equal.
> 
> ```
> $ion_symbol_table::{
>     symbols:["foo", "bar"]
> }
> 
> $ion_symbol_table::annotated::{
>     symbols:["foo", "bar"]
> }
> 
> $ion_symbol_table::annotated::{
>     symbols:abc::["foo", def::"bar"]
> }
> ```

This isn't quite right. Those system-level values aren't being directly tested for equivalence at all; we're actually comparing [the user-level symbol values](https://github.com/amzn/ion-tests/blob/034503caa56683f086edb1cc05a1b6207c1b18af/iontestdata/good/equivs/localSymbolTableWithAnnotations.ion#L11) that you've omitted.

As long as the `$10` and `$11` in each of those embedded streams resolve to the same symbols, we don't care what the serialized symbol tables looked like; we don't need to compare them to each other at all.

Comparing serialized symbol table tables in and of itself feels a bit dangerous. How would this code handle LST appends, where there are [multiple serialized `$ion_symbol_table` structs](https://github.com/amzn/ion-tests/blob/034503caa56683f086edb1cc05a1b6207c1b18af/iontestdata/good/equivs/localSymbolTableAppend.ion#L18-L39) that collectively result in an equivalent symbol table?

Reply:
Turns out that https://github.com/amzn/ion-go/pull/149 actually resolves all of this rendering my solution unnecessary. I removed all my logic for this problem from this PR and told Byron to remove `localSymbolTableWithAnnotations.ion` from the skip lists in his PR since his PR fixes this scenario.